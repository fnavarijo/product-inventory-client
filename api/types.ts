export * from '@/api/auth/types';
export * from '@/api/home/types';
export * from '@/api/products/types';
