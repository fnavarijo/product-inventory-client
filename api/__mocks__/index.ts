export const Products = {
  getAll: jest.fn(() => [{ id: 1 }, { id: 2 }, { id: 3 }]),
};
