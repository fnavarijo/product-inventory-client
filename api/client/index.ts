import { NuxtAxiosInstance } from '@nuxtjs/axios';

let axiosInstance: NuxtAxiosInstance;

export const initAxios = ({ axios }: { axios: NuxtAxiosInstance }) => {
  console.info('Axios instance set up');
  // @ts-ignore
  axiosInstance = axios;
};

export const getClient = () => ({
  axios: axiosInstance,
});
