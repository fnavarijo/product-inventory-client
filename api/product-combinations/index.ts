export * from '@/api/product-combinations/getAll';
export * from '@/api/product-combinations/getById';
