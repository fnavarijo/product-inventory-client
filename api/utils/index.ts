export * from '@/api/utils/resetProps';
export * from '@/api/utils/transformProduct';
export * from '@/api/utils/transformBy';
export * from '@/api/utils/renameProp';
