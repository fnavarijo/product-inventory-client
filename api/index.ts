import { initAxios } from './client';

export * as Auth from './auth';
export * as Home from './home';
export * as Products from './products';
export * as ProductCombinations from './product-combinations';

export {
  initAxios,
};
