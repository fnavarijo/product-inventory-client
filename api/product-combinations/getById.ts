import map from 'lodash/fp/map';
import get from 'lodash/fp/get';
import flow from 'lodash/fp/flow';
import pick from 'lodash/fp/pick';

import { getClient } from '@/api/client';
import { transformBy } from '@/api/utils';

import { SingleProductCombination } from '@/api/types';

export async function getById (id: string): Promise<SingleProductCombination> {
  const { data } = await getClient()
    .axios
    .get(`/product-combinations/${id}`);

  // TODO: I'd like to improve naming
  const pickProductBase = pick(['products', 'image']);
  const pickProduct = pick(['code', 'description', 'category', 'colors']);
  const pickColor = pick(['name', 'color']);
  const getImageUrl = get('url');
  const getCategoryName = get('name');

  const pickAllColors = map(pickColor);
  const filterColors = transformBy(pickAllColors)('colors');
  const filterCategoryName = transformBy(getCategoryName)('category');

  const pickAllProduct = map(flow([pickProduct, filterCategoryName, filterColors]));

  const filterProducts = transformBy(pickAllProduct)('products');
  const filterImage = transformBy(getImageUrl)('image');

  return flow([
    pickProductBase,
    filterProducts,
    filterImage,
  ])(data);
}
