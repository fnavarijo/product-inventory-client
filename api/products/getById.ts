import { getClient } from '@/api/client';
import { transformProduct } from '@/api/utils';

import { Product } from '@/api/types';

export async function getById (productId: string): Promise<Product> {
  const { data } = await getClient()
    .axios
    .get(`/products/${productId}`);

  return transformProduct(data);
}
