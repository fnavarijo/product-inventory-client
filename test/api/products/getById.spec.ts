import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { Products, initAxios } from '@/api';

import { mock } from '@/test/api/products/getById.mock';
import { NuxtAxiosInstance } from '@nuxtjs/axios';

describe('Products.getById function', () => {
  beforeAll(() => {
    const adapter = new MockAdapter(axios);

    adapter.onGet('/products/601a077f39aaa829dd99de3e').reply(200, mock);

    initAxios({ axios: axios as NuxtAxiosInstance });
  });

  it('should fetch and transform product', async () => {
    const product = await Products.getById('601a077f39aaa829dd99de3e');

    expect(product).toHaveProperty('id');
    expect(product).toHaveProperty('name');
    expect(product).toHaveProperty('category');
    expect(product).toHaveProperty('featuredImage');
    expect(product).toHaveProperty('isFeatured');
    expect(product).toHaveProperty('description');
    expect(product).toHaveProperty('images');
    expect(product).toHaveProperty('colors');
    expect(product).toHaveProperty('availableSizes');

    expect(product).toMatchSnapshot();
  });
});
