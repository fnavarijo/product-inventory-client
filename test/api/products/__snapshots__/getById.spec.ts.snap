// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products.getById function should fetch and transform product 1`] = `
Object {
  "availableSizes": Array [
    "S",
    "M",
  ],
  "category": "Accesorios - BeanBags",
  "colors": Array [
    Object {
      "color": "#000000",
      "name": "Negro",
    },
  ],
  "description": "Bean bag relleno de duroport para que pueda ser utilizado donde se desee debido a su forma elastica",
  "featuredImage": "https://res.cloudinary.com/dkvtpo8w1/image/upload/v1612318190/bean_bag_black_29a8280a9d.png",
  "id": "602a8e297703d2202bee1ceb",
  "images": Array [],
  "isFeatured": false,
  "name": "Bean Bag de cuero 2",
}
`;
